cli/lsp/diagnostics.rs

```diff
@@ -0,0 +1,261 @@
+            severity: Some(match d.category {
+              diagnostics::DiagnosticCategory::Error => {
+                lsp_types::DiagnosticSeverity::Error
+              }
+              diagnostics::DiagnosticCategory::Warning => {
+                lsp_types::DiagnosticSeverity::Warning
+              }
+              diagnostics::DiagnosticCategory::Suggestion => {
+                lsp_types::DiagnosticSeverity::Hint
+              }
+              diagnostics::DiagnosticCategory::Message => {
+                lsp_types::DiagnosticSeverity::Information
+              }
+            }),
+            code: Some(lsp_types::NumberOrString::Number(d.code as i32)),
+            code_description: None,
+            source: Some("deno-ts".to_string()),
+            message: get_diagnostic_message(d),
+            related_information: to_lsp_related_information(
+              &d.related_information,
+            ),
+            tags: match d.code {
+              // These are codes that indicate the variable is unused.
+              6133 | 6192 | 6196 => {
+                Some(vec![lsp_types::DiagnosticTag::Unnecessary])
+              }
+              _ => None,
+            },
+            data: None,
+          })
+        } else {
+          None
+        }
+      })
+      .collect(),
+  )
+}
+
+pub fn generate_ts_diagnostics(
+  state: &ServerStateSnapshot,
+  runtime: &mut JsRuntime,
+) -> Result<DiagnosticVec, AnyError> {
+  if !state.config.settings.enable {
+    return Ok(Vec::new());
+  }
+  let mut diagnostics = Vec::new();
+  let file_cache = state.file_cache.read().unwrap();
+  for (specifier, doc_data) in state.doc_data.iter() {
+    let file_id = file_cache.lookup(specifier).unwrap();
+    let version = doc_data.version;
+    let current_version = state.diagnostics.get_version(&file_id);
+    if version != current_version {
+      // TODO(@kitsonk): consider refactoring to get all diagnostics in one shot
+      // for a file.
+      let request_semantic_diagnostics =
+        tsc::RequestMethod::GetSemanticDiagnostics(specifier.clone());
+      let mut ts_diagnostics = ts_json_to_diagnostics(tsc::request(
+        runtime,
+        state,
+        request_semantic_diagnostics,
+      )?)?;
+      let request_suggestion_diagnostics =
+        tsc::RequestMethod::GetSuggestionDiagnostics(specifier.clone());
+      ts_diagnostics.append(&mut ts_json_to_diagnostics(tsc::request(
+        runtime,
+        state,
+        request_suggestion_diagnostics,
+      )?)?);
+      let request_syntactic_diagnostics =
+        tsc::RequestMethod::GetSyntacticDiagnostics(specifier.clone());
+      ts_diagnostics.append(&mut ts_json_to_diagnostics(tsc::request(
+        runtime,
+        state,
+        request_syntactic_diagnostics,
+      )?)?);
+      if !ts_diagnostics.is_empty() {
+        diagnostics.push((file_id, version, ts_diagnostics));
+      } else {
+        diagnostics.push((file_id, version, Vec::new()));
+      }
```

Comment:
Why is this conditional necessary?

Reply:
Because you have to send an empty array to potentially clear down any previous diagnostics that might have cleared. I realise that isn't obvious and will add a comment.

Reply:
I think his point is that if `ts_diagnostics` is empty, it's the same as `Vec::new()`

Reply:
🤦